File: Scripts/QueryScripts/UniqueQueries.pl
#!/usr/bin/perl -w
BEGIN { push @INC, '/Users/ricardocunha/Documents/FEUP/5ano/MSc Thesis/Thesis/Code/Scripts/Utils' }
require "sessionFileIndexes.pl";

use DB_File;
%queries = ();
tie(%queries, 'DB_File', "queries.dbfile", $DB_BTREE);

$numberQueries = 0;
$uniqueQueries = 0;

sub isQueryOnList{
    $query = shift;
    return exists($queries{$query});

}

sub processQuery{
    $keywords = shift;
    if(!isQueryOnList($keywords)){
        $queries{$keywords} = 1;
        $uniqueQueries++;
    } else {
        $queries{$keywords}++;
    }
}

sub printResults{
    print "unique>>>$uniqueQueries\n";
    print "total>>>$numberQueries\n";
    $ratio = sprintf("%.2f",($uniqueQueries/$numberQueries)*100);
    print "ratio>>>$ratio%\n";
    
    my $counter=0;
    my $top1000queries=0;
    foreach $query ( sort{$queries{$b} <=> $queries{$a}} keys %queries){
        if($counter==1000){
            last;
        }
        
        $top1000queries += $queries{$query};
        $counter++;
    }
    print "top1000>>>" . ($top1000queries/$numberQueries)*100;
    
}

sub main{
    while($_ = <STDIN>){
        @query = split(/>>>/, $_);
        
        $keywords = $query[$keywordsIndex];
        
        processQuery($keywords);
        $numberQueries++;
    }
    
    printResults();
}

main;
untie(%queries);

File: Scripts/QueryScripts/UniqueQueries.R
source("/Users/ricardocunha/Documents/FEUP/5ano/MSc\ Thesis/Thesis/Code/Scripts/Utils/mainBody.R")

uniqueQueries <- 0
totalQueries <- 0
ratio <- 0
top1000 <- 0

state <- 1
parseString <- function(str){
  aQuery = strsplit(str,separator)
  if(state == 1){
    uniqueQueries <<- aQuery[[1]][2]
    state <<- 2
  } else if(state == 2){
    totalQueries <<- aQuery[[1]][2]
    state <<- 3
  } else if(state == 3){
    ratio <<- aQuery[[1]][2]
    state <<- 4
  } else if(state == 4){
    top1000 <<- aQuery[[1]][2]
  }
}

printStats <- function(filename){
  write("UNIQUE QUERIES",file=filename,append=FALSE)
  write(paste("Number of unique queries: ",uniqueQueries,sep=" -> "),file=filename,append=TRUE)
  write(paste("Ratio: ",ratio,sep=" -> "),file=filename,append=TRUE)
  write(paste("Top 1000 unique queries account for(X) of total queries: ",top1000,sep=" -> "),file=filename,append=TRUE)
  
}

main <- function(){
  readFromStdin()
  filename <- paste(outputPath,"UniqueQueries.txt", sep = "/")
  printStats(filename)
}

main()